File: Algorithms/Easy/169.Majority-Element/MajorityElement.go
package easy

// Time complexity: O(n)
// Space complexity: O(n)
func majorityElement_1(nums []int) int {
	counter := make(map[int]int)
	for _, num := range nums {
		counter[num]++
	}

	majElem, countElem := 0, 0
	for elem, count := range counter {
		if count > countElem {
			majElem = elem
			countElem = count
		}
	}

	return majElem
}

// Time complexity: O(n)
// Space complexity: O(1)
//
// Boyer–Moore majority vote algorithm
func majorityElement_2(nums []int) int {
	majElem, counter := 0, 0
	for _, num := range nums {
		if counter == 0 {
			majElem = num
			counter = 1
		} else if majElem == num {
			counter++
		} else {
			counter--
		}
	}

	return majElem
}

File: Algorithms/Easy/169.Majority-Element/MajorityElement.py
class Solution:
    # Time complexity: O(n)
    # Space complexity: O(n)
    def majorityElement_1(self, nums: list[int]) -> int:
        counter = {}
        for num in nums:
            counter[num] = counter.get(num, 0) + 1

        maxEl, maxCount = 0, 0
        for elem, count in counter.items():
            if count > maxCount:
                maxEl = elem
                maxCount = count

        return maxEl

    # Time complexity: O(n)
    # Space complexity: O(1)
    # Boyer–Moore majority vote algorithm
    def majorityElement_2(self, nums: list[int]) -> int:
        counter, res = 0, 0
        for num in nums:
            if counter == 0:
                res = num
                counter = 1
            elif num == res:
                counter += 1
            else:
                counter -= 1

        return res
